### DISCUSSION TITLE
Obfuscating an Uno WASM app with Dotfuscator results in only Uno icon and no errors

### DISCUSSION THREAD

**Discussion by andrewkdci** [unoplatform/uno#11985] [Q&A] [answered]
Hello all,

I am trying to obfuscate the WASM version of a proof of concept app that I wrote. I am using Dotfuscator 6.4 to do the obfuscation on the .clr after publishing the WASM app. I am running the app on Firefox (64-bit, v108.0.2) and on Ubuntu Desktop 20.04.

The app loads and runs fine unobfuscated. However when I obfuscate it, all I get is the Uno logo after the loading screen completes. I cannot figure out how to see any errors. I even modified my App.xaml.cs to put a try-catch around the entire OnLaunched event, which will show an error screen with a stack trace in the event of an exception during startup. I've tested this with injected exceptions and it works as expected when unobfuscated. However when I run obfuscated I do not get the error screen. Just the Uno icon.

So my first question is: Where can I see any errors Uno might have in loading a WASM app? I've tried opening up the developer options in my browser (F12 in Firefox), but I don't see anything useful in there.

To at least prove out that Uno WASM can be obfuscated, I created a very small test app with only a single View Model and a single Model. The MainView page just shows a value from the Model as exposed via the ViewModel. This runs fine when obfuscated, so I know that obfuscating an Uno WASM app is possible. To add to the mystery, though, going back to my PoC, when I change the obfuscation settings to exclude renaming and control-flow for EVERYTHING, it STILL does not load properly. So I imagine there must be something in my PoC code that is causing Dotfuscator to mess something up when obfuscating-but-not-really-obfuscating the .clr.

So my second question is: Does anyone have any experience or knowledge of caveats that should be taken into account when obfuscating a .clr from a published WASM app?

I appreciate any thoughts/ideas.

**Comment by andrewkdci** [comment]
I seem to have found the culprit, although I'm not sure why. It looks like my IoC container is to blame. I've already noticed that Uno WASM doesn't work with Autofac. So I switched to TinyIoC. That worked with unobfuscated code.

Now, it looks like that even when I just register types with the TinyIoC container, it will break my obfuscated code. I don't even need to resolve the types. When I comment out all UI controls so that my main page is just a bunch of tabs, but am still registering types with the IoC container, then it doesn't load the WASM. When I comment out the registration of types, then the WASM does load.

I'll verify this by adding TinyIoC to the hello world that I got working as obfuscated, and will update here what I find.

**Comment by andrewkdci** [comment]
Welp, I don't know what's going on. 

So first of all, adding TinyIoC to my other hello world project did not break it. So it doesn't seem like it's intrinsic to that IoC container.

Then, in my PoC I've whittled down the code to where it's not even referencing any IoC container, and it still breaks. This is my Dependency Service, with all the code commented out referring to any IoC container:

    internal enum DependencyInjectionFramework
    {
        Autofac,
        TinyIoC
    }

    internal static class MyDependencyService
    {
        //private static Lazy<IServiceLocator> _serviceLocatorFactory;
        ////private static Lazy<ContainerBuilder> _builderFactory = new Lazy<ContainerBuilder>(() => new ContainerBuilder());

        //public static IServiceLocator ServiceLocator => _serviceLocatorFactory.Value;
        public static IServiceLocator ServiceLocator => new MockServiceLocator();

        public static DependencyInjectionFramework DIFramework { get; set; }

        static MyDependencyService()
        {
            //_serviceLocatorFactory = new Lazy<IServiceLocator>(() =>
            //{
            //    //switch (DIFramework)
            //    //{
            //    //    case DependencyInjectionFramework.TinyIoC:
            //    var tinyIoCcontainer = TinyIoCContainer.Current;
            //    return new TinyIoCServiceLocator(tinyIoCcontainer);

            //    //    case DependencyInjectionFramework.Autofac:
            //    //    default:
            //    //        var autofacContainer = _builderFactory.Value.Build();
            //    //        return new AutofacServiceLocator(autofacContainer);
            //    //}
            //});
        }

        /// <summary>
        /// Override the default instance of the service locator.
        /// </summary>
        /// <param name="serviceLocator"></param>
        public static void OverrideLocatorInstance(IServiceLocator serviceLocator)
        {
            //_serviceLocatorFactory = new Lazy<IServiceLocator>(() => serviceLocator);
        }

        //public static void RegisterTypes(Action<ContainerBuilder> registrationAction)
        //{
        //    //registrationAction?.Invoke(_builderFactory.Value);
        //}

        //public static void RegisterTypes(Action<TinyIoCContainer> registrationAction)
        //{
        //    registrationAction?.Invoke(TinyIoCContainer.Current);
        //}
    }

As you can see, almost all of it is commented out.

And then by adding JUST this line in the App.xaml.cs.OnLaunched:

    MyDependencyService.DIFramework = DependencyInjectionFramework.TinyIoC;

it will no longer work obfuscated. Without this line, then obfuscated works (albeit with no real functionality because I don't have any VMs or Models).

I'm tearing my hair out here trying to identify why this is failing once obfuscated.

**Comment by andrewkdci** [comment]
BTW, in case anyone is interested, here is my Dotfuscator config that I'm using:

```
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<!--This config file was generated by Dotfuscator. Please use the Dotfuscator Config Editor to modify.-->
<!DOCTYPE dotfuscator SYSTEM "http://www.preemptive.com/dotfuscator/dtd/dotfuscator_v2.5.dtd">
<dotfuscator version="2.3">
  <global>
    <input_management>manual</input_management>
    <option>monocompat</option>
  </global>
  <input>
    <loadpaths>
      <file dir="${configdir}/bin/Release/net6.0" />
    </loadpaths>
    <asmlist>
      <inputassembly refid="97021910-dd0a-4131-b651-c1d121573b96">
        <option>honoroas</option>
        <option>stripoa</option>
        <file dir="C:\Users\user\source\repos\Feasibility\Feasibility.Uno.Wasm\bin\Release\net6.0\publish\package_4e74784aaec3b524c13a20ee9614f80280594dc2\managed" name="Feasibility.Uno.Wasm.clr" />
      </inputassembly>
    </asmlist>
  </input>
  <output>
    <file dir="C:\Users\user\Documents\Documents_Local\VMShared\NginxStaging" />
  </output>
  <renaming>
    <option>xmlserialization</option>
    <excludelist>
      <assembly>
        <file dir="" name="Feasibility.Uno.Wasm.clr" />
      </assembly>
    </excludelist>
    <mapping>
      <mapoutput overwrite="true">
        <file dir="${configdir}\DotfuscatorReports\${configuration}\${platform}" name="Renaming.xml" />
      </mapoutput>
    </mapping>
    <referencerulelist>
      <referencerule rulekey="{6655B10A-FD58-462d-8D4F-5B1316DFF0FF}" />
      <referencerule rulekey="{7D9C8B02-2383-420f-8740-A9760394C2C1}" />
      <referencerule rulekey="{229FD6F8-5BCC-427b-8F72-A7A413ECDF1A}" />
      <referencerule rulekey="{2B7E7C8C-A39A-4db8-9DFC-6AFD38509061}" />
      <referencerule rulekey="{494EA3BA-B947-44B5-BEE8-A11CC85AAF9B}" />
      <referencerule rulekey="{89769974-93E9-4e71-8D92-BE70E855ACFC}" />
      <referencerule rulekey="{4D81E604-A545-4631-8B6D-C3735F793F80}" />
      <referencerule rulekey="{62bd3899-7d53-4336-8ca2-4e5dbae187d5}" />
    </referencerulelist>
  </renaming>
  <controlflow level="high">
    <excludelist>
      <assembly>
        <file dir="" name="Feasibility.Uno.Wasm.clr" />
      </assembly>
    </excludelist>
  </controlflow>
  <removal>
    <option>disable</option>
    <removalreport overwrite="true">
      <file dir="${configdir}\DotfuscatorReports\${configuration}\${platform}" name="Removal.xml" />
    </removalreport>
    <referencerulelist />
  </removal>
  <sos />
  <smartobfuscation>
    <smartobfuscationreport verbosity="all" overwrite="true">
      <file dir="${configdir}\DotfuscatorReports\${configuration}\${platform}" name="SmartObfuscation.xml" />
    </smartobfuscationreport>
  </smartobfuscation>
</dotfuscator>
```

Salient points to consider:
1. I am excluding the entire assembly from renaming and control flow. I believe that this means that the output is theoretically the same as the input.
2. The advanced settings for the obfuscation are default except that I set "Use only Mono-compatible transforms" = true.
3. The rest is default. No string encryption, no removal, etc.

**Comment by andrewkdci** [comment]
So I removed all IoC code and am building the dependency tree manually. With that out of the way, I finally got a more useful (hopefully) error message to show up on the error screen:

![image](https://user-images.githubusercontent.com/130710084/233734558-573ebb1c-dca0-4348-adff-f87d2e199eb2.png)

(Company and project name elided in order to maintain at least some degree of confidentiality)

**Comment by andrewkdci** [comment]
So at this point I have two issues:

1. Using an IoC container (I have so far tried Autofac and TinyIoC), I only get the blank page with the Uno logo after loading.
2. When I remove the IoC container and manually build my dependency tree, I get the error in the previous comment.

**Reply by darenm** [reply, marked as answer]
Have you tried updating `LinkerConfig.json` to include your assemblies? I'm wondering whether the obfuscation coupled with IoC is causing essential code to be trimmed. [Using the IL Linker](https://platform.uno/docs/articles/features/using-il-linker-webassembly.html)

**Reply by andrewkdci** [reply]
Hi @darenm,

Thanks for this. It certainly is a HUGE step in the right direction!

I added the WasmShellILLinkerEnabled = false tag to my WASM csproj file, as recommended to determine whether the linker is to blame. The result was that it at least stopped hanging at the Uno logo and is now displaying the error and stack trace of my error page. However now it looks like I'm having type resolution issues inside IoC container itself:

Resolution error when I use Autofac:
![DI resolution error - Autofac](https://user-images.githubusercontent.com/130710084/233745394-ef0f5bd3-6ac8-4538-ab96-8f295dbe9947.PNG)

Resolution error when I use TinyIoC:
![DI resolution error - TinyIoC](https://user-images.githubusercontent.com/130710084/233745397-18224a16-eabe-432c-8521-4618efe80f5f.PNG)

There is no reason why the IoC container shouldn't be able to resolve these types. It still runs fine without obfuscation. Interestingly, though, the different DI frameworks fail at different spots. TinyIoC seems to be unable to find MainViewModel when resolving from the IoC container (ax.ar maps to my ServiceLocator.Resolve method), while Autofac seems to have some strange error when resolving further down the resolution chain (SystemAccessViewModel contains a reference to h [ISystemAccess], which contains a reference to k [IFileAccess]). This may be a red herring, tho...

So, problem not solved completely, but at least we now know that the linker was causing troubles, so removing that was at least a big step in the right direction.

**Reply by andrewkdci** [reply]
Marking this reply by @darenm as the answer since it did fix the issue with only the Uno logo displaying when running an obfuscated app. I will make another question that is now more focused on getting DI to work in an obfuscated app.

Thank you, @darenm !
